File: thanh123456_AcceptedCodeVNSPOJ/QMAX.pas
const
   fin='';fon='';
var
   n,m:longint;
   a,t:array [1..trunc(1e6)] of int64;
procedure nhap;
   begin
   assign(input,fin);reset(input);
   assign(output,fon);rewrite(output);
   readln(n,m);
   end;
procedure taomang;
   var
      i,u,v,x:longint;
   begin
   for i:=1 to m do
      begin
      readln(u,v,x);
      inc(a[u],x);
      dec(a[v+1],x);
      end;
   for i:=1 to n do a[i]:=a[i-1]+a[i];
   end;
function max(a,b:int64):int64;
   begin
   if a>b then exit(a)
   else exit(b);
   end;
procedure taocay(i,l,r:longint);
   var
      mid:longint;
   begin
   if l=r then
      begin
      t[i]:=a[l];
      exit;
      end;
   mid:=(l+r) div 2;
   taocay(i*2,l,mid);
   taocay(i*2+1,mid+1,r);
   t[i]:=max(t[i*2],t[i*2+1]);
   end;
function tinh(i,l,r,u,v:longint):int64;
   var
      left,right:int64;
      mid:longint;
   begin
   if (r<u) or (l>v) then exit(0)
      else
         if (l>=u) and (r<=v) then exit(t[i])
            else
               begin
               mid:=(l+r) div 2;
               left:=tinh(i*2,l,mid,u,v);
               right:=tinh(i*2+1,mid+1,r,u,v);
               exit(max(left,right));
               end;
   end;
procedure xuat;
   var
      i,q,u,v:longint;
   begin
   readln(q);
   for i:=1 to q do
      begin
      readln(u,v);
      writeln(tinh(1,1,n,u,v));
      end;
   close(input);
   close(output);
   end;
begin
nhap;
taomang;
taocay(1,1,n);
xuat;
end.

File: thanh123456_AcceptedCodeVNSPOJ/QMAX.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1e5 + 100;
const int oo = (1 << 31) - 1;
int n,m;
struct ds
{
    int lazy,res;
}it[N * 4];
void nhap()
{
    scanf("%d %d",&n,&m);
}
void cn(int i,int l,int r)
{
    if (it[i].lazy == 0) {return;}
    it[i].res+=it[i].lazy;
    if (l != r)
    {
        it[i * 2].lazy += it[i].lazy;
        it[i * 2 + 1].lazy += it[i].lazy;
    }
    it[i].lazy = 0;
}
void update(int i,int l,int r,int u,int v,int x)
{
    cn(i,l,r);
    if (l > v || r < u) {return;}
    if (l >= u && r <= v)
    {
        it[i].lazy += x;
        cn(i,l,r);
        return;
    }
    int mid = (l + r) >> 1;
    update(i*2,l,mid,u,v,x);
    update(i*2+1,mid+1,r,u,v,x);
    it[i].res = max(it[i * 2].res,it[i * 2 + 1].res);
}
int get(int i,int l,int r,int u,int v)
{
    cn(i,l,r);
    if (l > v || r < u) {return -oo;}
    if (l >= u && r <= v) {return it[i].res;}
    int mid = (l + r) >> 1;
    int left = get(i * 2,l,mid,u,v);
    int right = get(i * 2 +1,mid + 1,r,u,v);
    return max(left,right);
}
void solve()
{
    for (int i = 1; i <= n; i++) {update(1,1,n,i,i,0);}
    for (int i = 1; i <= m; i++)
    {
        int u,v,x;
        scanf("%d %d %d",&u,&v,&x);
        update(1,1,n,u,v,x);
    }
    int q;
    scanf("%d",&q);
    for (int i = 1; i <= q; i++)
    {
        int u,v;
        scanf("%d %d",&u,&v);
        printf("%d\n",get(1,1,n,u,v));
    }
}
int main()
{
    nhap();
    solve();
}
